src/components/cylc/graph/Graph.vue

```diff
@@ -0,0 +1,631 @@
+      if (!this.freeze) {
+        this.loading = true
+        const elements = {
+          nodes: [],
+          edges: []
+        }
+        const filtered = workflows.filter((workflow) => workflow.name === this.workflowName)
+        if (filtered.length !== 0) {
+          const workflow = filtered[0]
+          if (Object.hasOwnProperty.call(workflow, 'nodesEdges')) {
+            if (Object.hasOwnProperty.call(workflow.nodesEdges, 'edges')) {
+              workflow.nodesEdges.edges.forEach((edge) => {
+                elements.edges.push({
+                  data: edge
+                })
+              })
+            }
+            if (Object.hasOwnProperty.call(workflow.nodesEdges, 'nodes')) {
+              workflow.nodesEdges.nodes.forEach((node) => {
+                elements.nodes.push({
+                  data: node
+                })
+              })
+            }
+          }
+        }
+        // clear existing data
+        this.graphData.edges.length = 0
+        this.graphData.nodes.length = 0
+        // assign new data (need to assign a new one, so that the watcher(deep:false) triggers off)
+        this.graphData = elements
+      }
+    },
+
+    // --- Begin initialization functions
+
+    /**
+     * Register the extensions used by Cytoscape for this component.
+     */
+    registerExtensions () {
+      if (typeof cytoscape('core', 'navigator') !== 'function') {
+        navigator(cytoscape)
+      }
+      if (typeof cytoscape('core', 'panzoom') !== 'function') {
+        panzoom(cytoscape)
+      }
+      if (typeof cytoscape('core', 'undoRedo') !== 'function') {
+        undoRedo(cytoscape)
+      }
+      if (typeof cytoscape('core', 'popper') !== 'function') {
+        popper(cytoscape)
+      }
+      if (typeof cytoscape('core', 'nodeHtmlLabel') !== 'function') {
+        nodeHtmlLabel(cytoscape)
+      }
+    },
+
+    /**
+     * Runs the current layout.
+     * @param {cytoscape} instance - the cytoscape instance
+     * @param {*} [layoutOptions=null] - the layout options
+     * @see https://js.cytoscape.org/#cy.layout
+     */
+    runLayout (instance, layoutOptions = null) {
+      instance
+        .elements()
+        .layout(layoutOptions || this.layoutOptions)
+        .run()
+    },
+
+    /**
+     * Sets up the undo-redo plugin in the instance.
+     * @param {cytoscape} instance - the cytoscape instance
+     */
+    setupUndoRedo (instance) {
+      const undoRedoOptions = {
+        isDebug: false, // Debug mode for console messages
+        actions: {}, // actions to be added
+        undoableDrag: true, // Whether dragging nodes are undoable can be a function as well
+        stackSizeLimit: undefined, // Size limit of undo stack, note that the size of redo stack cannot exceed size of undo stack
+        ready: function () {
+          // callback when undo-redo is ready
+          this.loading = false
+        }
+      }
+      instance.undoRedo(undoRedoOptions)
+    },
```

Comment:
Not quite sure undo-redo is working... but left it anyway. Lumino also has some key listeners...

Reply:
What is is meant to do?

Reply:
After you modify the position of nodes, for example, pressing `CTRL + Z` should undo that, and move your node back to its original position.

Demo: https://rawgit.com/iVis-at-Bilkent/cytoscape.js-undo-redo/master/demo.html